docs_nnx/guides/pytree.md

```diff
@@ -8,60 +8,159 @@ jupytext:
+
+```{code-cell} ipython3
+class Foo:
+  def __init__(self):
+    self.a = 1
+    self.b = 2
+    self.c = "hi"
+
+def flatten_foo(foo: Foo):
+  nodes = [foo.a, foo.b]  # sequence of nodes
+  static = (foo.c,) # hashable & equatable structure
+  return nodes, static
+
+def unflatten_foo(static, nodes):
+  foo = object.__new__(Foo)  # create uninitialized instance
+  foo.a = nodes[0]
+  foo.b = nodes[1]
+  foo.c = static[0]
+  return foo
+
+jax.tree_util.register_pytree_node(Foo, flatten_foo, unflatten_foo)
+
+foo = Foo()
+leaves, treedef = jax.tree.flatten(foo)
+print(f"leaves = {leaves}")
+print(f"treedef = {treedef}")
+```
+
+Notice that `'hi'` does not appear in the leaves because `c` is defined as static, but you can see it as part of the `PyTreeDef` structure.
+
++++
+
+## nnx.Pytree
+In general it would be cumbersome for users to manually register the pytree definition for every type they create. To automate this process NNX provides the `nnx.Pytree` base type that offers a simple API: users annotate attributes using either `nnx.static` or `nnx.data`, and Pytree will register some flatten and unflatten functions that will take the annotations into account. The `nnx.data` and `nnx.static` annotations must only be assigned to `Pytree` attributes directly.
 
 ```{code-cell} ipython3
 class Linear(nnx.Pytree):
   def __init__(self, din: int, dout: int):
-    self.din, self.dout = din, dout
-    self.w = jnp.ones((din, dout))
-    self.b = jnp.zeros((dout,))
+    self.din = nnx.static(din)
+    self.dout = nnx.static(dout)
+    self.w = nnx.data(jnp.ones((din, dout)))
+    self.b = nnx.data(jnp.zeros((dout,)))
 
 class MLP(nnx.Pytree):
   def __init__(self, num_layers, dim):
-    self.num_layers = num_layers
-    self.layers = nnx.List([Linear(dim, dim) for _ in range(num_layers)])
+    self.num_layers = nnx.static(num_layers)
+    self.layers = nnx.data([
+      Linear(dim, dim) for _ in range(num_layers)
+    ])
 
 pytree = MLP(num_layers=2, dim=1)
+
+def pytree_structure(pytree, title='pytree structure'):
+  print(f"{title}:")
+  path_leaves, treedef = jax.tree.flatten_with_path(pytree)
+  for path, value in path_leaves:
+    print(f" - pytree{jax.tree_util.keystr(path)} = {value!r}")
+
 pytree_structure(pytree)
 ```
 
-### Attribute Annotations
+As you can see above, only the `data` paths appear in the leaves. However, its very verbose to have to define `static` and `data` for each attribute so `Pytree` has sensible defaults so you can mostly just remove them from them and it will work.
```

Comment:
nit: `However, its very verbose to have to define `static` and `data` for each attribute, so `Pytree` has sensible defaults. You can remove most of them and it will just work:`